JSON API improvements

G10h4ck opened the pull request.

G10h4ck committed:
Improve debugabbility of JSON API

JSON API server uses standardized error handling based on
  std::error_condition for a few method, this improve debugabbility both
  from library size and from API client side
JSON API auth token are now more flexible, now the only proibition is using
  colon character ':' in user name
Implemented TO_JSON operation for std::error_condition
Avoid unkown serial job handling code duplication via fatalUnknownSerialJob
Usability improvement in t_RsLogger
Disable compilation of now unused is_alphanumeric

G10h4ck committed:
Fix global shutdown call from JSON API

G10h4ck committed:
Delete experimental deadcode

csoler commented on libretroshare/src/jsonapi/jsonapi.cpp:
```diff
@@ -127,18 +129,16 @@ JsonApiServer::corsOptionsHeaders =
-	for(uint32_t i=0; i < tkLen && colonCounter < 2; ++i)
-		if(clear_token[i] == ':') { ++colonCounter; colonIndex = i; }
-		else if(!librs::util::is_alphanumeric(clear_token[i])) return false;
+	for(uint32_t i=0; i < tkLen; ++i)
```
This is ambiguous. What happens if the token is "user:passwd:passwd" ? then the user will be "user:passwd". That means a passwd cannot contain ":".  In summary, I think that the passwd should be required to be alphanumeric. Otherwise you need to send it in base64 in which case the passwd can be anything and the parsing can be made non ambiguous by ensuiring that the token contains exactly one ":" and that the encoded passwd is what remains after ":".

G10h4ck replied:
If the token is like "somethingA:somethinchB:somethingC" then user is "somethingA" and password is "somethinchB:somethingC"

G10h4ck replied:
In fact the token is transferred as a base64 encoded string, RFC7617 has specific rules, IMHO the most relevant parts are:

- For the user-id, recipients MUST support all characters defined in the "UsernameCasePreserved" profile defined in Section 3.3 of [RFC7613], with the exception of the colon (":") character.

- For the password, recipients MUST support all characters defined in the "OpaqueString" profile defined in Section 4.2 of [RFC7613].

- Constructs the user-pass by concatenating the user-id, a single colon (":") character, and the password

Basically in the token what is before the first ':' is to be considered username what is before is to be considered password, and both user and password con be non alphanumeric and not even ASCII

G10h4ck replied:
The line is now `if(clear_token[i] == ':') { colonIndex = i; break; }`

csoler commented on libretroshare/src/serialiser/rstypeserializer.h:
```diff
@@ -909,6 +910,27 @@ struct RsTypeSerializer
 	                       t_RsTlvList<TLV_CLASS,TLV_TYPE>& member,
 	                       RsJson& jDoc );
 
+	[[noreturn]] static void fatalUnknownSerialJob(int j)
```
what's this????

G10h4ck replied:
https://en.cppreference.com/w/cpp/language/attributes/noreturn  it tells that the function won't return, aka if the program get there it will be terminated like when you call exit(...)

csoler commented on libretroshare/src/util/rsstring.h:
```diff
@@ -30,10 +30,10 @@ namespace librs {
 		bool ConvertUtf16ToUtf8(const std::wstring& source, std::string& dest);
-
+#if 0
 		bool is_alphanumeric(char c) ;
```
has these two been moved somewhere else?

G10h4ck replied:
They are not used anymore so I kept them under `#ifdef 0` if we need them in the future.

G10h4ck committed:
Fixup token parsing

If the token is like "somethingA:somethinchB:somethingC" then user is
"somethingA" and password is "somethinchB:somethingC"

G10h4ck committed:
Solve race condition in JSON API restart

Improve atomic access to restbed service
Protect JSON API restart against bursts
Fix JSON API error condition enum registration
Provide ostream helper for error_condition
Provide optional forced thread cancel for debugging purpose, disabled by
  default at compile time define RS_THREAD_FORCE_STOP to enable it
Avoid double fullstop in retroshare-gui json api apply button
